src/java.base/share/classes/java/io/FileOutputStream.java

```diff
@@ -313,23 +314,41 @@ public void write(int b) throws IOException {
         boolean append = FD_ACCESS.getAppend(fd);
         long comp = Blocker.begin();
         try {
-            write(b, append);
+            write0(b, append);
         } finally {
             Blocker.end(comp);
         }
     }
 
+    private native int writeBytes0(byte[] b, int off, int len, boolean append)
+        throws IOException;
+
     /**
      * Writes a sub array as a sequence of bytes.
      * @param b the data to be written
      * @param off the start offset in the data
      * @param len the number of bytes that are written
-     * @param append {@code true} to first advance the position to the
-     *     end of file
      * @throws    IOException If an I/O error has occurred.
      */
-    private native void writeBytes(byte[] b, int off, int len, boolean append)
-        throws IOException;
+    private void writeBytes(byte[] b, int off, int len)
+        throws IOException
+    {
+        Objects.checkFromIndexSize(off, len, b.length);
+        boolean append = FD_ACCESS.getAppend(fd);
+        int pos = off;
+        int remaining = len;
+        long comp = Blocker.begin();
+        try {
+            do {
+                int size = Math.min(remaining, 1572864);
+                int n = writeBytes0(b, pos, size, append);
+                pos += n;
+                remaining -= n;
+            } while (remaining > 0);
```

Comment:
Long standing behavior is that write is a no-open with len is 0, meaning you can do while here (no need to change it to do-while). Same thing in RAF.

Reply:
So changed in cdb8455ccbd1db49d272e867b16a14eb72b41a7b.